Hyphenate @RestHeader sourceName by default

jjaferson opened the pull request:
Hyphenate @RestHeader sourceName by default

Resolves https://github.com/quarkusio/quarkus/issues/13665

jjaferson committed:
Hyphenate @RestHeader name by default

gastaldi commented on independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/StringUtil.java:
```diff
@@ -0,0 +1,104 @@
+        };
+    }
+
+    /**
+     * @deprecated Use {@link String#join} instead.
+     * @param delim delimiter
+     * @param it iterator
+     * @return the joined string
+     */
+    @Deprecated
+    public static String join(final String delim, final Iterator<String> it) {
+        final StringBuilder b = new StringBuilder();
+        if (it.hasNext()) {
+            b.append(it.next());
+            while (it.hasNext()) {
+                b.append(delim);
+                b.append(it.next());
+            }
+        }
+        return b.toString();
+    }
```
Remove it anyway? I don't think anyone is using it TBH

jjaferson replied:
I added this because [`String.join`](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-) receives an `Iterable` and `camelHumpsIterator` method return an `Iterator` but I think it's better to cast the return from `camelHumpsIterator` to an iterable and use `String.join`.

gastaldi commented on independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/StringUtil.java:
```diff
@@ -0,0 +1,89 @@
+        };
+    }
+
+    public static <T> Iterable<T> iteratorToIterable(Iterator<T> iterator) {
+        return () -> iterator;
+    }
```
Is this method really necessary? And needs to be public?

jjaferson replied:
not really we can cast the return of `camelHumpsIterator` inside of `hyphenate` as you suggest below.

gastaldi commented on independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/StringUtil.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public static String hyphenate(final String orig) {
+        return String.join("-", iteratorToIterable(camelHumpsIterator(orig)));
```
```suggestion
        return String.join("-", () -> camelHumpsIterator(orig));
```

jjaferson replied:
I think we need the return type to be `Iterable<String>`

```   
public static String hyphenate(final String orig) {
  return String.join("-", (Iterable<String>) () -> camelHumpsIterator(orig));
}
```

gastaldi reviewed (approved):
LGTM

gastaldi commented:
I'll let @geoand have the final say on this PR

geoand commented:
Thanks for this!

I have a question though: what happens if the user actually sets a value in the `@RestParam` annotation?

gastaldi commented:
> I have a question though: what happens if the user actually sets a value in the `@RestParam` annotation?

According to https://github.com/quarkusio/quarkus/pull/16003/files#diff-81713f62adc649d447ab44049899a4edd826be9be784b1aa3b654b4a22febf44R825 I think it will use what's defined there.

@jjaferson can we avoid calling the `StringUtil.hyphenate` when a value is set? That operation seems costly and unnecessary in this case

jjaferson commented:
> e avoid calling the `StringUtil.hyphenate` when a value i

Good catch, that makes seems, I added a check to verify if the rest param is set before hyphenating it.

jjaferson commented:
> Thanks for this!
> 
> I have a question though: what happens if the user actually sets a value in the `@RestParam` annotation?

no problem, that would be hyphenated again but now I added a check to verify if the restParam was set.

geoand commented:
> > Thanks for this!
> > I have a question though: what happens if the user actually sets a value in the `@RestParam` annotation?
> 
> no problem, that would be hyphenated again but now I added a check to verify if the restParam was set.

We would also need a test for this behavior as well

gastaldi commented on independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java:
```diff
@@ -822,6 +822,8 @@ public PARAM extractParameterInfo(ClassInfo currentClassInfo, ClassInfo actualEn
             builder.setType(ParameterType.HEADER);
             convertible = true;
         } else if (restHeaderParam != null) {
+            if (restHeaderParam.value() == null)
+                sourceName = StringUtil.hyphenate(sourceName);
             builder.setName(valueOrDefault(restHeaderParam.value(), sourceName));
```
Don't overwrite sourceName, this is a pain to debug. Better do this instead:

```suggestion
            if (restHeaderParam.value() == null) {
            	builder.setName(StringUtil.hyphenate(sourceName));
            } else {
            	builder.setName(restHeaderParam.value().asString());
            }
```

jjaferson replied:
I see, as we won't be using `valueOrDefault` I'll add a check for empty values as well. 

```            
if (restHeaderParam.value() == null || restHeaderParam.value().asString().isEmpty()) {
                builder.setName(StringUtil.hyphenate(sourceName));
}
```

jjaferson commented:
> > > Thanks for this!
> > > I have a question though: what happens if the user actually sets a value in the `@RestParam` annotation?
> > 
> > 
> > no problem, that would be hyphenated again but now I added a check to verify if the restParam was set.
> 
> We would also need a test for this behavior as well

added tests for all the cases - when the @restParam isn't specified and when it's empty

geoand reviewed (approved):
Thanks!

FroMage commented:
Huh, I was meaning to review this PR, but never got notified, so I didn't know it was filed. Don't we get `/CC` on PRs?

FroMage commented:
I mean, your reviews were good, I wouldn't have done it differently, but I'm surprised I didn't get notified.

geoand commented:
There is no notification from the bot on PRs unfortunately

jjaferson commented:
> Huh, I was meaning to review this PR, but never got notified, so I didn't know it was filed. Don't we get `/CC` on PRs?

Sorry @FroMage I should have pinged you on the PR, I thought you would be notified when I added the reference to the issue you created...

FroMage commented:
> There is no notification from the bot on PRs unfortunately

Is this on purpose, or due to a limitation, or just that we're missing a bot action?

geoand commented:
Likely the latter.

Something maybe that @SaumyaSingh1 can work on if @gsmet agrees?

SaumyaSingh1 commented:
Sure :)

FroMage commented:
Yeah, great idea!
